### PR TITLE
issue #171: introduce property-based testing for pure core domain logic

### LINKED ISSUE #171: Introduce property-based testing for pure core domain logic
# Introduce property-based testing for pure core domain logic

## Summary

Introduce **property-based testing (PBT)** in the `core` module to strengthen the unit tests of our pure domain logic. PBT lets us assert **invariants and properties** over a large, randomized input space instead of a handful of hand-picked examples, surfacing edge cases (rounding, boundaries, overflow) that example-based tests routinely miss.

The scope is deliberately narrow: PBT applies **only to pure, deterministic domain logic** (Value Objects and pure functions). It **complements** the existing example/table-based tests — it does not replace them.

## Motivation

The Value Objects in `core/.../domain/commons` (`Percentage`, `Temperature`, `AbsoluteHumidity`, `ThermoHygroData`) carry clear invariants but are currently covered only by a few `@TestFactory`/`dynamicTest` example cases. These are valuable as golden/regression tests but explore a tiny slice of the input space. PBT closes that gap by checking that the invariants hold for thousands of generated inputs, with automatic shrinking to a minimal counterexample on failure.

### In scope
- **Pure Value Objects and pure functions** in `core`, e.g.:
  - `Percentage` — value always in `[0, 1]`; `ofHundreds` round-trip; rejection of out-of-range inputs.
  - `Temperature` — `average()` properties (bounds, single-element identity, ordering).
  - `AbsoluteHumidity` — invariants only (non-negativity, monotonicity), **not** recomputation of the formula.
  - `ThermoHygroData` — construction/aggregation invariants.

### Out of scope
- **Mock-based application services** (e.g. `CreateAreaServiceTest`, `*ServiceTest` driven by mockk). PBT over mocked collaborators adds complexity without meaningful value.
- The `persistence`, `providers`, and `app` modules (integration/adapter tests). Possible future extension, not part of this issue.

### 1. Library: `kotest-property`
We already use Kotest (assertions) on a JUnit 5 runner. Add the `kotest-property` module and use it **inside the existing JUnit 5 tests** via `checkAll` / `forAll` (e.g. `runBlocking { checkAll(Arb.percentage()) { ... } }`). **No migration of the test runner to Kotest Spec classes.**

### 2. Properties/invariants, never re-implement the production formula
PBT must assert **properties** — bounds, monotonicity, round-trip, symmetry, idempotence — and must **not** re-implement the production computation in the test (that would be a self-confirming tautology). Exact computed values (e.g. the Magnus-formula cases in `AbsoluteHumidityTest`) remain covered by the existing example/table tests, which stay as-is.

### 3. Generators (`Arb<T>`) live in `core/testFixtures`; object mothers are untouched
Add a small set of reusable domain generators (`Arb.temperature()`, `Arb.percentage()`, …) in `core/src/testFixtures`. The existing object mothers (`BaseMothers.kt`, etc.) are **not** restructured: tests simply feed the generated value-under-test into the existing mother factory parameters (which already accept values with random defaults), e.g.:

```kotlin
checkAll(Arb.temperature()) { t ->
    val data = aRandomThermoHygroData(temperature = t)
    // assert invariants on `data`
}
```

This requires `kotest-property` as a `testFixturesImplementation` dependency in `core`. Generators also give us shrinking, which the current `Random`-based mothers do not.

### 4. Iterations and seed
- **1000 iterations** as the default for property tests.
- **Fixed seed** (via `PropTestConfig(seed = …)` or `PropertyTesting.defaultSeed`) for reproducible CI failures. Kotest prints the failing seed to reproduce locally.

### 5. Mutation testing (pitest): exclude PBT classes
pitest re-runs covering tests per mutant; a 1000-iteration PBT would both blow up build time **and** make the mutation score non-deterministic if inputs vary between runs.

Decision: **mark PBT tests with a dedicated annotation and exclude them from pitest by JUnit tag.** Introduce a self-documenting meta-annotation that bundles a JUnit 5 `@Tag`:

```kotlin
@Target(AnnotationTarget.CLASS, AnnotationTarget.FUNCTION)
@Retention(AnnotationRetention.RUNTIME)
@Tag("pbt")
annotation class PropertyBasedTest
```

and exclude that group in the pitest block:

```kotlin
// core/build.gradle.kts — pitest block
excludedGroups = listOf("pbt")
```

This is preferred over a class-name convention: it is explicit at the test site, survives renames, can mark individual methods within a mixed class, and the same tag can be used to filter PBT in the normal `test` task if ever needed. `excludedGroups` is supported by the `pitest-junit5` plugin (already in use) and maps onto JUnit 5 `@Tag`s.

The existing example/table tests remain the mutation oracle (they are designed for that), so the `70` mutation threshold is unaffected and the build stays fast. PBT runs in the normal `test` task only.

## Acceptance criteria

- [ ] `kotest-property` added to `gradle/libs.versions.toml` and wired into `core` (`testImplementation` + `testFixturesImplementation`).
- [ ] Reusable `Arb<T>` domain generators added under `core/src/testFixtures` (`Arb.percentage()`, `Arb.temperature()`, and others as needed).
- [ ] A `@PropertyBasedTest` meta-annotation (bundling `@Tag("pbt")`) is added and applied to all PBT classes/methods.
- [ ] Property tests added for the in-scope Value Objects, asserting invariants (not re-implemented formulas).
- [ ] Each property test runs **1000 iterations** with a **fixed seed**.
- [ ] Existing example/table tests are kept unchanged.
- [ ] Object mothers are unchanged in structure (only value-under-test is substituted).
- [ ] PBT are excluded from pitest via `excludedGroups = listOf("pbt")`; `./gradlew :core:pitest` still passes the `70` threshold and build time is not materially increased.
- [ ] `./gradlew :core:test` is green.

## Out-of-scope / follow-ups
- Extending PBT to `persistence` / `providers` adapters.
- Revisiting whether selected PBTs should contribute to the mutation score (would require throttled iterations under pitest).

## Notes
- Verify the exact Kotest system-property keys for iteration count / seed during implementation.
- Verify that `excludedGroups` correctly excludes `@Tag("pbt")` tests under `pitest-junit5` 1.2.1.
- Remember the project conventions: no `CHANGELOG.md` entry is required (test-only addition), and this is a structural/test change — TDD workflow does not apply.

### PR DESCRIPTION
**Pull request by AGRFesta**

### CODE REVIEW

File: core/src/test/kotlin/org/agrfesta/sh/api/domain/commons/TemperaturePropertyTest.kt [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+                Temperature.of(t.value) shouldBe t
+            }
+        }
+    }
+
+    @Test
+    fun `compareTo is consistent with the underlying value`() {
+        runBlocking {
+            checkAll(pbtConfig, Arb.temperature(), Arb.temperature()) { a, b ->
+                a.compareTo(b).sign shouldBe a.value.compareTo(b.value).sign
```
[the remark this answers is not in the document]
Comment (AGRFesta) [reply core/src/test/kotlin/org/agrfesta/sh/api/domain/commons/TemperaturePropertyTest.kt:32]: @gemini-code-assist this is a false positive. `sign` here is the `kotlin.math.sign` extension property on `Int` (imported at the top of the file: `import kotlin.math.sign`), which returns `-1/0/1`. Since `compareTo` returns `Int`, `compareTo(...).sign` compiles and behaves as intended — the module builds and `:core:test` is green. Keeping the current form, as it reads more clearly than nesting a second `compareTo(0)`.

File: core/src/test/kotlin/org/agrfesta/sh/api/domain/commons/TemperaturePropertyTest.kt [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        }
+    }
+
+    @Test
+    fun `average of a non-empty collection lies within min and max`() {
+        runBlocking {
+            checkAll(pbtConfig, Arb.list(Arb.temperature(), 1..20)) { temps ->
+                val avg = temps.average()!!
+                (avg in temps.min()..temps.max()) shouldBe true
+            }
```
[the remark this answers is not in the document]
Comment (AGRFesta) [reply core/src/test/kotlin/org/agrfesta/sh/api/domain/commons/TemperaturePropertyTest.kt:61]: @gemini-code-assist this guidance is outdated for our Kotlin version (2.1.21). `Iterable.min()`/`max()` were deprecated back in 1.4 but reintroduced as non-nullable in Kotlin 1.7, so they are not deprecated here and compile without warnings. Switching to `minOrNull()!!`/`maxOrNull()!!` would only add redundant `!!` operators on a collection that is already guaranteed non-empty (`1..20`). Keeping the current form.
